### PR TITLE
add support for unmappedTargetPolicy

### LINKED ISSUE #132: Use `unmappedTargetPolicy` in the unmapped target properties inspection

### PR DESCRIPTION
**Pull request by hduelme**
I added support for `unmappedTargetPolicy`. The policy will be read from the class annotation `@Mapper` first. If not configured on class level, the property of the mapper-config-class will be used. Otherwise the fallback policy warn is used.

To directly match the compiler results, `unmappedTargetPolicy = ERROR` results in Errors getting reported instead of warnings. 
If set to `unmappedTargetPolicy = IGNORE` all warnings for unmapped  targets are suppressed in the current class.

Fixes #132

### CODE REVIEW

File: src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java [hunk 1]
```diff
@@ -479,4 +482,78 @@ private static Stream<PsiClass> findReferencedMappersOfMapperConfig(PsiAnnotatio
...
+                                                                 @NotNull ReportingPolicy fallback ) {
+        PsiClass containingClass = method.getContainingClass();
+        if (containingClass == null) {
+            return fallback;
+        }
+        return getReportingPolicyFromClass( containingClass, attributeName, fallback );
+    }
+
+    @NotNull
+    public static ReportingPolicy getReportingPolicyFromClass( @NotNull PsiClass containingClass,
```
Comment (filiphr) [inline src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java:497]: This doesn't need to be public I think

File: src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java [hunk 2]
```diff
@@ -479,4 +482,78 @@ private static Stream<PsiClass> findReferencedMappersOfMapperConfig(PsiAnnotatio
         return findReferencedMappers( mapperConfigAnnotation );
     }
 
+    @NotNull
+    public static ReportingPolicy getReportingPolicyFromMethode( @NotNull PsiMethod method,
```
Comment (filiphr) [inline src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java:486]: I would suggest to rename this to `getUnmappedTargetPolicy` and only pass the `mappingMethod`. We are using this method in only one place so we can do the fallback in the implementation of the method and the use of `unmappedTargetPolicy`.

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyConfigClassTest.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Tests if unmappedTargetPolicy could be read from mapper config class.
+ * @author hduelme
+ */
+public class UnmappedTargetPropertiesInspectionReportPolicyConfigClassTest extends BaseInspectionTest {
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
Comment (filiphr) [inline src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyConfigClassTest.java:24]: We don't need to override this. Let's use the default for our tests (Java 1.8)

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyClassLevelFirstTest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Tests if unmappedTargetPolicy is read from class first. Class level annotation should overwrite config class values.
+ * @author hduelme
+ */
+public class UnmappedTargetPropertiesInspectionReportPolicyClassLevelFirstTest extends BaseInspectionTest {
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
Comment (filiphr) [inline src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyClassLevelFirstTest.java:24]: We don't need to override this. Let's use the default for our tests (Java 1.8)

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyErrorTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Tests if unmappedTargetPolicy set to ERROR, results in reported errors instead of warnings.
+ * @author hduelme
+ */
+public class UnmappedTargetPropertiesInspectionReportPolicyErrorTest extends BaseInspectionTest {
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
Comment (filiphr) [inline src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyErrorTest.java:24]: We don't need to override this. Let's use the default for our tests (Java 1.8)

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyIgnoreTest.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Tests if unmappedTargetPolicy set to IGNORE, suppress all unmapped warnings
+ * @author hduelme
+ */
+public class UnmappedTargetPropertiesInspectionReportPolicyIgnoreTest extends BaseInspectionTest {
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
Comment (filiphr) [inline src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyIgnoreTest.java:24]: We don't need to override this. Let's use the default for our tests (Java 1.8)

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyStaticImportTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Tests if unmappedTargetPolicy could be read, if static import is used.
+ * @author hduelme
+ */
+public class UnmappedTargetPropertiesInspectionReportPolicyStaticImportTest extends BaseInspectionTest {
+
+    @Override
+    protected LanguageLevel getLanguageLevel() {
+        return LanguageLevel.JDK_1_7;
```
Comment (filiphr) [inline src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyStaticImportTest.java:24]: We don't need to override this. Let's use the default for our tests (Java 1.8)

**Review by filiphr — CHANGES_REQUESTED** [review]
Thanks @hduelme.

I've left some comments inline.

Apart from those can we perhaps do the following:

* remove the `_after` tests. I think that it is enough to only validate that the warnings / errors are there.
* Can we please also add some tests where you have `unmappedTargetPolicy = WARN" on the mapper, but `unmappedTargetPolicy = ERROR` on the mapper config?
* Can we please also take `BeanMapping#unmappedTargetPolicy` into consideration as well?

File: src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java (hunk 1 above)
Comment (hduelme) [reply src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java]: Changed to `private`

File: src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java (hunk 2 above)
Comment (hduelme) [reply src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java]: Renamed to getUnmappedTargetPolicy

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyClassLevelFirstTest.java (hunk 4 above)
Comment (hduelme) [reply src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyClassLevelFirstTest.java]: Removed

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyConfigClassTest.java (hunk 3 above)
Comment (hduelme) [reply src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyConfigClassTest.java]: Removed

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyIgnoreTest.java (hunk 6 above)
Comment (hduelme) [reply src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyIgnoreTest.java]: Removed

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyStaticImportTest.java (hunk 7 above)
Comment (hduelme) [reply src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyStaticImportTest.java]: Removed

File: src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyErrorTest.java (hunk 5 above)
Comment (hduelme) [reply src/test/java/org/mapstruct/intellij/inspection/UnmappedTargetPropertiesInspectionReportPolicyErrorTest.java]: Removed

**Comment by hduelme** [comment]
@filiphr I added support for `@BeanMapping`. I also renamed the methods as suggested and removed the not needed parameters. Furthermore, I have added more tests, which cover different possible combinations of overwriting policy.

**Comment by filiphr** [comment]
I also updated the description of the inspection to include the fact that the unmapped target policy is taken into consideration

**Comment by hduelme** [comment]
@filiphr thanks for the feedback. I personally was not sure, if it would be better to have all tests in the same place. But I a agree. It looks cleaner.

**Pull request merged**